lib/internal/webstreams/writablestream.js

```diff
@@ -0,0 +1,1270 @@
+  switch (stream[kState].state) {
+    case 'writable':
+      if (!writableStreamCloseQueuedOrInFlight(stream) &&
+          stream[kState].backpressure) {
+        writer[kState].ready = createDeferredPromise();
+      } else {
+        writer[kState].ready = {
+          promise: PromiseResolve(),
+          resolve: undefined,
+          reject: undefined,
+        };
+      }
+      setClosedPromiseToNewPromise();
+      break;
+    case 'erroring':
+      writer[kState].ready = {
+        promise: PromiseReject(stream[kState].storedError),
+        resolve: undefined,
+        reject: undefined,
+      };
+      setPromiseHandled(writer[kState].ready.promise);
+      setClosedPromiseToNewPromise();
+      break;
+    case 'closed':
+      writer[kState].ready = {
+        promise: PromiseResolve(),
+        resolve: undefined,
+        reject: undefined,
+      };
+      writer[kState].close = {
+        promise: PromiseResolve(),
+        resolve: undefined,
+        reject: undefined,
+      };
+      break;
+    default:
+      writer[kState].ready = {
+        promise: PromiseReject(stream[kState].storedError),
+        resolve: undefined,
+        reject: undefined,
+      };
+      writer[kState].close = {
+        promise: PromiseReject(stream[kState].storedError),
+        resolve: undefined,
+        reject: undefined,
+      };
+      setPromiseHandled(writer[kState].ready.promise);
+      setPromiseHandled(writer[kState].close.promise);
+  }
+
+  function setClosedPromiseToNewPromise() {
+    writer[kState].close = createDeferredPromise();
+  }
+}
+
+function writableStreamAbort(stream, reason) {
+  const {
+    state,
+  } = stream[kState];
+  if (state === 'closed' || state === 'errored')
+    return PromiseResolve();
+
+  if (stream[kState].pendingAbortRequest.abort.promise !== undefined)
+    return stream[kState].pendingAbortRequest.abort.promise;
+
+  assert(state === 'writable' || state === 'erroring');
+
+  let wasAlreadyErroring = false;
+  if (state === 'erroring') {
+    wasAlreadyErroring = true;
+    reason = undefined;
+  }
+
+  const abort = createDeferredPromise();
+
+  stream[kState].pendingAbortRequest = {
+    abort,
+    reason,
+    wasAlreadyErroring,
+  };
+
+  if (!wasAlreadyErroring)
+    writableStreamStartErroring(stream, reason);
+
+  return abort.promise;
+}
+
+function writableStreamClose(stream) {
+  const {
+    state,
+    writer,
+    backpressure,
+    controller,
+  } = stream[kState];
+  if (state === 'closed' || state === 'errored') {
+    return PromiseReject(
+      new ERR_INVALID_STATE.TypeError('WritableStream is closed'));
+  }
+  assert(state === 'writable' || state === 'erroring');
+  assert(!writableStreamCloseQueuedOrInFlight(stream));
+  stream[kState].closeRequest = createDeferredPromise();
+  const { promise } = stream[kState].closeRequest;
+  if (writer !== undefined && backpressure && state === 'writable')
+    writer[kState].ready.resolve?.();
```

Comment:
is this user code being invoked? if so, it needs to be try/catched so a rejected promise can be returned if needed

Reply:
No, these are deferreds created in [`setupWritableStreamDefaultWriter`](https://github.com/nodejs/node/blob/721b8cdbee12fa6ad5700dcbc643bae733322f1b/lib/internal/webstreams/writablestream.js#L508). While `writer[kState].ready.promise` is exposed to author code as `writer.ready`, any fulfillment/rejection handlers attached to that promise are run in a separate microtask.